File: backend/app/schemas/schedule.py
"""Schedule request/response schemas (#830).

Three vocabularies the rest of the feature is built from — `SessionIntent`,
`Discipline`, `Commitment` — declared once here as `Literal`s so the API, the
store and (later) the coach's structured output all coerce against the same set.
`ScreenPointer` is the precedent: a closed `Literal` plus `extra="forbid"` means
an off-contract value fails at the boundary rather than reaching a column.

`SpacingRule` follows the `ProposedActionRequest` idiom deliberately: one flat
model whose per-kind arguments are optional fields, with a `model_validator`
enforcing the required set for each kind. A tagged union would read more neatly
in isolation, but this shape is what the house already validates LLM-supplied
arguments with, and this rule set will be LLM-supplied.
"""

from datetime import date, datetime
from typing import Any, Dict, List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.schemas.coach_context import VolumeMetricComparison

# --- the three axes --------------------------------------------------------

SessionIntent = Literal["rest", "easy", "long", "quality", "strength"]
Discipline = Literal["run", "walk", "bike", "strength", "row", "other"]
Commitment = Literal["committed", "suggested"]

# Derived, never stored: see models/planned_session.py.
Placement = Literal["pinned", "window", "week"]
SessionStatus = Literal["upcoming", "done", "missed", "dismissed"]

RuleKind = Literal[
    "rest_day_after",
    "no_intent_day_before",
    "min_days_between",
    "preferred_days",
    "max_sessions_per_day",
]


# --- rules -----------------------------------------------------------------


class SpacingRule(BaseModel):
    """One spacing constraint the week must satisfy.

    The kinds are a closed vocabulary with a pure predicate each
    (`services/schedule/rules.py`), which is what makes a violation DETECTABLE
    rather than merely readable. The four rules the design's flexible week is
    made of all express here:

        "No quality run the day before the long run"
            -> no_intent_day_before(before_intent=quality, target_intent=long)
        "No heavy legs the day before a quality run"
            -> no_intent_day_before(before_intent=strength, target_intent=quality)
        "A full rest day after the long run"
            -> rest_day_after(intent=long)
        "Long run needs a free morning — Sat or Sun"
            -> preferred_days(intent=long, weekdays=[5, 6])

    `label` is the coach's OWN prose — carried so the coach can phrase a rule in
    its own terms — but it is not what the runner is told the rule IS (#844: a
    live label promised "or an easy walk" against a `rest_day_after` predicate
    that forbids exactly that, and the runner planned from the label rather than
    the predicate). The runner-facing STATEMENT of what a rule enforces is
    derived in code from `kind` + its arguments (`services/schedule/rule_text.py`
    `describe_rule`) on the READ side only — see `SpacingRuleRead` below — so it
    can never claim more or less than the PREDICATE (`rules.py`) actually checks.
    """

    model_config = ConfigDict(extra="forbid")

    kind: RuleKind
    label: str = Field(min_length=1, max_length=200)
    source: Literal["coach", "runner"] = "coach"

    # rest_day_after / preferred_days
    intent: Optional[SessionIntent] = None
    # no_intent_day_before
    before_intent: Optional[SessionIntent] = None
    target_intent: Optional[SessionIntent] = None
    # min_days_between
    intent_a: Optional[SessionIntent] = None
    intent_b: Optional[SessionIntent] = None
    days: Optional[int] = Field(default=None, ge=1, le=7)
    # preferred_days — 0 = Monday .. 6 = Sunday, matching services/weeks.py
    weekdays: Optional[List[int]] = Field(default=None, min_length=1, max_length=7)
    # max_sessions_per_day
    count: Optional[int] = Field(default=None, ge=1, le=5)

    @field_validator("weekdays")
    @classmethod
    def _weekdays_in_range(cls, v: Optional[List[int]]) -> Optional[List[int]]:
        if v is not None and any(d < 0 or d > 6 for d in v):
            raise ValueError("weekdays must be 0 (Monday) through 6 (Sunday)")
        return v

    @model_validator(mode="after")
    def _validate_shape(self) -> "SpacingRule":
        if self.kind == "rest_day_after" and self.intent is None:
            raise ValueError("rest_day_after requires intent")
        if self.kind == "no_intent_day_before" and (
            self.before_intent is None or self.target_intent is None
        ):
            raise ValueError(
                "no_intent_day_before requires before_intent and target_intent"
            )
        if self.kind == "min_days_between" and (
            self.intent_a is None or self.intent_b is None or self.days is None
        ):
            raise ValueError("min_days_between requires intent_a, intent_b and days")
        if self.kind == "preferred_days" and (
            self.intent is None or not self.weekdays
        ):
            raise ValueError("preferred_days requires intent and weekdays")
        if self.kind == "max_sessions_per_day" and self.count is None:
            raise ValueError("max_sessions_per_day requires count")
        return self


class RuleViolation(BaseModel):
    """A rule the current placement cannot satisfy.

    `statement` (#844) is the derived runner-facing text — see
    `SpacingRuleRead`/`rule_text.describe_rule` — and is what a violation should
    be READ as breaking. `label` is kept unchanged (the coach's own prose) so a
    caller that already renders it is not silently repointed at different text;
    it is additive, not a repurposing of what `label` means.
    """

    model_config = ConfigDict(extra="forbid")

    kind: RuleKind
    label: str
    statement: str
    detail: str


class SpacingRuleRead(SpacingRule):
    """`SpacingRule` plus its derived runner-facing `statement` (#844).

    Read-only and additive: `SpacingRuleRead` is never the type of a stored rule
    or of the LLM's drafted output (`draft_contract.DraftedPlan.rules` stays
    `List[SpacingRule]`), so nothing about what is persisted or what the coach is
    asked for changes. `statement` is computed at read time
    (`rule_text.describe_rule`) from `kind` + arguments — never from `label` — so
    it can never promise more than the rule's predicate (`rules.py`) enforces.
    """

    statement: str


# --- the horizon's week shape ----------------------------------------------


class PlannedWeekShape(BaseModel):
    """One week of the horizon, as SHAPE rather than sessions.

    Concrete for ~3 weeks, shape only beyond (a settled design decision). A week
    that also has `planned_sessions` rows reads as planned; a week with only this
    reads as sketched. Nothing stores which — see `services/schedule/horizon.py`.
    """

    model_config = ConfigDict(extra="forbid")

    week_start: date
    phase: Optional[str] = Field(default=None, max_length=60)
    target_running_distance_m: Optional[float] = Field(default=None, ge=0)
    target_effort_score: Optional[float] = Field(default=None, ge=0)
    # discipline -> share of the week's load, 0..1. Shares, not absolutes, so a
    # mix cannot contradict the total it is a mix of.
    discipline_mix: Dict[str, float] = Field(default_factory=dict)
    intent_mix: Dict[str, float] = Field(default_factory=dict)


# --- sessions --------------------------------------------------------------


class SessionStructure(BaseModel):
    """The interval shape `workout_matching.match_planned_to_detected` expects.

    Its exact keys, unchanged: that function has compared a plan to detected reps
    since the beginning and has never once been handed one, because
    `_extract_planned_workout` is a placeholder that returns None.
    """

    model_config = ConfigDict(extra="forbid")

    reps_planned: int = Field(ge=1, le=60)
    rep_distance_m: Optional[float] = Field(default=None, gt=0)
    rest_s: Optional[float] = Field(default=None, ge=0)


class PlannedSessionRead(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    window_start: date
    window_end: date
    # Derived from the window; see models/planned_session.py.
    placement: Placement
    # `max(window_start, today) .. window_end`, or None once the window has
    # passed. The stored window never moves.
    effective_window_start: Optional[date] = None
    effective_window_end: Optional[date] = None
    # True once time has eaten into the window — what the design shows as
    # "window narrowed from Thu-Sat". Carried rather than left for the client to
    # re-derive, so the stored-vs-effective comparison is made in one place.
    has_narrowed: bool = False

    intent: SessionIntent
    discipline: Discipline
    commitment: Commitment
    status: SessionStatus

    title: str
    detail: Optional[str] = None
    target_distance_m: Optional[float] = None
    target_duration_s: Optional[int] = None
    target_effort_score: Optional[float] = None
    structure: Optional[Dict[str, Any]] = None

    completed_at: Optional[datetime] = None
    completed_activity_id: Optional[UUID] = None
    completion_source: Optional[str] = None
    dismissed_at: Optional[datetime] = None


# --- logged actuals --------------------------------------------------------


class LoggedActivityRead(BaseModel):
    """What the runner actually did this week, from the one fact stream.

    Projected from `activity_facts`, never re-queried: the schedule must not be a
    second opinion about a week's totals.
    """

    model_config = ConfigDict(extra="forbid")

    activity_id: Optional[UUID] = None
    local_date: date
    activity_type: str
    discipline: Discipline
    distance_m: float
    moving_time_s: int
    effort_score: float


# --- the week read ---------------------------------------------------------


class DisciplineLoad(BaseModel):
    """One discipline's slice of the week, planned and logged.

    Sized by `effort_score` because it is the only unit that sums honestly across
    a gym session and a turbo ride; km cannot, since strength and an indoor bike
    have no distance.
    """

    model_config = ConfigDict(extra="forbid")

    discipline: Discipline
    planned_effort_score: float = 0.0
    logged_effort_score: float = 0.0
    planned_sessions: int = 0
    logged_sessions: int = 0


class WeekHeadline(BaseModel):
    """Running km leads; load carries the disciplines km cannot measure.

    A settled decision: km is how a runner thinks about running and running is
    the priority sport, so the headline is run distance — with the discipline mix
    beneath it sized by load so strength, bike and row stay visible as progress.
    """

    model_config = ConfigDict(extra="forbid")

    planned_running_distance_m: Optional[float] = None
    logged_running_distance_m: float = 0.0
    planned_sessions: int = 0
    done_sessions: int = 0


class RunningVsNorm(BaseModel):
    """This week's RUNNING against the runner's own typical running week.

    The free-mode gauge's input. It exists because the headline is running km
    while the all-activity norm can be three times that for a runner who walks a
    lot — two different quantities in one glance. The gauge now measures the same
    thing the big number does.
    """

    model_config = ConfigDict(extra="forbid")

    typical_weekly_distance_m: float
    current_distance_m: float
    pct_vs_norm: float
    direction: str
    deadband_pct: float


class ScheduleWeekRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    week_start: date
    week_end: date
    is_current_week: bool
    # Whether an ACTIVE plan row exists. Note for the UI: free mode is "no
    # COMMITTED sessions" (`headline.planned_sessions == 0`), which is the state
    # the design describes — a plan carrying only suggestions is still free.
    # Render the free week off the headline, not off this flag.
    has_plan: bool
    plan_id: Optional[UUID] = None

    headline: WeekHeadline
    sessions: List[PlannedSessionRead] = Field(default_factory=list)
    logged: List[LoggedActivityRead] = Field(default_factory=list)
    by_discipline: List[DisciplineLoad] = Field(default_factory=list)
    rules: List[SpacingRuleRead] = Field(default_factory=list)
    violations: List[RuleViolation] = Field(default_factory=list)

    # The runner's own typical week, straight from the existing volume builder —
    # `norm_weekly` is ALL-ACTIVITY by that definition, with `current_runs`
    # carried alongside, so "typical" means exactly what it means on Trends.
    # Present only for the current week, where "as of today" is meaningful.
    norm: Optional[List[VolumeMetricComparison]] = None
    # The runs-only read, for the free-mode gauge. `norm` above stays as the
    # all-activity view the Trends page shows; this is the one that matches the
    # running-km headline it sits under.
    running_norm: Optional[RunningVsNorm] = None


# --- the horizon read ------------------------------------------------------


class HorizonWeek(BaseModel):
    model_config = ConfigDict(extra="forbid")

    week_start: date
    phase: Optional[str] = None
    # True when the week has real sessions; False when it is shape only. Derived
    # from what is actually there, so it cannot claim more than the plan holds.
    planned: bool
    # The four states a week can be in (#842). `planned`/`sketched` are the same
    # committed-session / week-shape derivation `planned` above already reads.
    # The other two are what used to be one byte-identical `else` branch: an
    # `empty` week is a genuine GAP the plan's own span covers but says nothing
    # about, while `beyond_plan` is a week past the plan's last covered week —
    # the plan never sketched it, so it must not wear the same "shape only, not
    # written yet" claim a real sketched week earns. `planned == (coverage ==
    # "planned")` always holds; see `services/schedule/horizon.py`.
    coverage: Literal["planned", "sketched", "empty", "beyond_plan"]
    is_current: bool
    running_distance_m: Optional[float] = None
    effort_score: Optional[float] = None
    discipline_mix: Dict[str, float] = Field(default_factory=dict)
    intent_mix: Dict[str, float] = Field(default_factory=dict)


class GoalRaceRead(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    name: str
    race_date: date
    distance_m: float
    priority: str


class ScheduleHorizonRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    weeks: List[HorizonWeek] = Field(default_factory=list)
    races: List[GoalRaceRead] = Field(default_factory=list)
    has_plan: bool
    # The largest weekly load in the window; bar lengths are true proportions of
    # it, so the ramp reads honestly instead of every bar looking maxed out.
    peak_effort_score: Optional[float] = None


# --- goal race writes ------------------------------------------------------


class GoalRaceCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=200)
    race_date: date
    distance_m: float = Field(gt=0, le=1_000_000)
    priority: Literal["A", "B", "C"] = "A"


# --- drafting --------------------------------------------------------------


class DraftStatusRead(BaseModel):
    """Where the runner's most recent plan stands.

    `status` is None when they have never had one. `drafting` is deliberately
    invisible to the week read, so the previous plan (or free mode) keeps serving
    while a new one is written — a runner asking for a new plan never loses the
    one they are following mid-week.
    """

    model_config = ConfigDict(extra="forbid")

    status: Optional[Literal["drafting", "active", "superseded", "failed"]] = None
    plan_id: Optional[UUID] = None
    generated_at: Optional[datetime] = None
    # Runner-facing. The validator's own failure text is internal and stays in the
    # log; what the runner is owed is a plain sentence.
    message: str

File: frontend/lib/types/schedule.ts
// #830: the schedule — what the runner is doing next. Mirrors backend
// app/schemas/schedule.py.
//
// Three independent axes describe a planned session: PLACEMENT (where in the
// week it sits), COMMITMENT (agreed to, or merely suggested) and DISCIPLINE
// (which sport). Intent is the coaching read — an easy bike and an easy run are
// the same stimulus — so intent carries the colour and discipline is named
// beside it, never given a competing colour of its own.
//
// Placement is DERIVED server-side from a stored inclusive window; there is no
// mode the runner picks. `effective_window_*` is that window narrowed to today
// onward, so a floating session's days shrink as the week passes while the
// stored window never moves.

import type { VolumeDirection, VolumeMetricVsNorm } from "./volume";

export type SessionIntent = "rest" | "easy" | "long" | "quality" | "strength";
export type Discipline =
  | "run"
  | "walk"
  | "bike"
  | "strength"
  | "row"
  | "other";
export type Commitment = "committed" | "suggested";
export type Placement = "pinned" | "window" | "week";
export type SessionStatus = "upcoming" | "done" | "missed" | "dismissed";

export type RuleKind =
  | "rest_day_after"
  | "no_intent_day_before"
  | "min_days_between"
  | "preferred_days"
  | "max_sessions_per_day";

export interface SpacingRule {
  kind: RuleKind;
  label: string;
  source: "coach" | "runner";
  intent?: SessionIntent | null;
  before_intent?: SessionIntent | null;
  target_intent?: SessionIntent | null;
  intent_a?: SessionIntent | null;
  intent_b?: SessionIntent | null;
  days?: number | null;
  weekdays?: number[] | null;
  count?: number | null;
}

// #844: the runner-facing rule TEXT the API actually returns. `statement` is
// derived server-side from `kind` + arguments — the only text guaranteed to
// match what the rule's predicate enforces. `label` is the coach's own prose,
// carried alongside as a subordinate note; a runner following `label` used to
// be able to trigger the very rule it described (a live label promised "or an
// easy walk" against a predicate that forbade exactly that).
export interface SpacingRuleRead extends SpacingRule {
  statement: string;
}

export interface RuleViolation {
  kind: RuleKind;
  label: string;
  statement: string;
  detail: string;
}

export interface PlannedSession {
  id: string;
  window_start: string;
  window_end: string;
  placement: Placement;
  effective_window_start: string | null;
  effective_window_end: string | null;
  has_narrowed: boolean;
  intent: SessionIntent;
  discipline: Discipline;
  commitment: Commitment;
  status: SessionStatus;
  title: string;
  detail: string | null;
  target_distance_m: number | null;
  target_duration_s: number | null;
  target_effort_score: number | null;
  structure: Record<string, number> | null;
  // How far this session goes, decided server-side by
  // `services/schedule/planned_distance.py` — the same value the week's
  // headline is summed from (#887). 0 means the session states no distance;
  // a duration is never converted into one.
  planned_distance_m: number;
  completed_at: string | null;
  completed_activity_id: string | null;
  completion_source: string | null;
  dismissed_at: string | null;
}

export interface LoggedActivity {
  activity_id: string | null;
  local_date: string;
  activity_type: string;
  discipline: Discipline;
  distance_m: number;
  moving_time_s: number;
  effort_score: number;
}

// Sized by effort_score, the only unit that sums across a gym session and a
// turbo ride; km cannot, because strength and an indoor bike have no distance.
export interface DisciplineLoad {
  discipline: Discipline;
  planned_effort_score: number;
  logged_effort_score: number;
  planned_sessions: number;
  logged_sessions: number;
}

export interface WeekHeadline {
  // null in free mode — there is no target to report.
  planned_running_distance_m: number | null;
  logged_running_distance_m: number;
  planned_sessions: number;
  done_sessions: number;
}

// The runs-only vs-norm read, which is what the free-mode headline is actually
// made of. `norm` below stays the all-activity Trends view; this one measures
// the same quantity the big running-km number does, so the gauge under it is
// not a second unit stacked on the first.
//
// `typical_weekly_distance_m` is the WHOLE week's typical, while
// `current_distance_m` is the week so far — the server pro-rates the norm by
// days elapsed before computing `pct_vs_norm`, so the percentage is fair on a
// Tuesday and the caption can still quote a full typical week.
export interface RunningVsNorm {
  typical_weekly_distance_m: number;
  current_distance_m: number;
  pct_vs_norm: number;
  direction: VolumeDirection;
  deadband_pct: number;
}

export interface ScheduleWeek {
  week_start: string;
  week_end: string;
  is_current_week: boolean;
  // Whether an ACTIVE plan row exists. NOT the free-mode signal: a plan holding
  // only suggestions is still a free week. Render free mode off
  // `headline.planned_sessions === 0`.
  has_plan: boolean;
  plan_id: string | null;
  headline: WeekHeadline;
  sessions: PlannedSession[];
  logged: LoggedActivity[];
  by_discipline: DisciplineLoad[];
  rules: SpacingRuleRead[];
  violations: RuleViolation[];
  // The runner's own typical week, from the same builder Trends uses. Present
  // only for the current week, where "as of today" means anything.
  norm: VolumeMetricVsNorm[] | null;
  // The runs-only read for the free-mode gauge. Present only for the current
  // week, and only once there is enough history to say what typical is.
  running_norm: RunningVsNorm | null;
}

export interface DraftStatus {
  status: "drafting" | "active" | "superseded" | "failed" | null;
  plan_id: string | null;
  generated_at: string | null;
  message: string;
}

// #857: the plan the runner was training to before this one.
//
// Always an object: "you have no earlier plan" is an answer with its own
// sentence, not a 404. `restorable` is the SERVER's verdict: the refusal has a
// reason the runner is owed (a plan whose horizon has passed would leave them
// with nothing planned), and a client re-deriving it from the dates would
// eventually derive it differently from the endpoint that enforces it.
export interface PreviousPlan {
  plan_id: string | null;
  // When it stopped being current, and when its thinking was written. Two
  // different facts: a plan replaced this morning can be one drafted in June.
  superseded_at: string | null;
  generated_at: string | null;
  horizon_end: string | null;
  // Sessions of that plan that still lie ahead: whether going back to it would
  // actually give the runner anything.
  sessions_ahead: number;
  restorable: boolean;
  message: string;
}

// A is the race the block is built towards; B and C are races run along the
// way. The read keeps `priority` as a plain string because the stored column is
// one — only the WRITE is closed to the three the API accepts.
export type RacePriority = "A" | "B" | "C";

export interface GoalRace {
  id: string;
  name: string;
  race_date: string;
  distance_m: number;
  priority: string;
}

export interface GoalRaceCreate {
  name: string;
  race_date: string;
  distance_m: number;
  priority: RacePriority;
}

// --- the horizon ------------------------------------------------------------
//
// The block's SHAPE rather than its sessions: concrete for about three weeks,
// sketched beyond. `planned` is derived server-side from what is actually
// there, so it can never claim a week holds sessions it does not.
//
// A week the plan says nothing about still arrives, with a null load and empty
// mixes, so the horizon stays a continuous run of weeks and a gap reads as a
// gap rather than as a shorter block.

// The four states a week can be in (#842). `planned`/`sketched` mirror the
// `planned` boolean below; `empty` is a genuine interior GAP the plan's own
// span covers but says nothing about, while `beyond_plan` is a week past the
// plan's last covered week — the plan never had a chance to sketch it, so it
// must not read as "shape only, not written yet" the way a real sketched week
// does. With no plan at all, every week is `empty`; there is nothing to be
// beyond.
export type HorizonCoverage = "planned" | "sketched" | "empty" | "beyond_plan";

export interface HorizonWeek {
  week_start: string;
  phase: string | null;
  // True when the week holds committed sessions; false when it is shape only.
  // Always equal to `coverage === "planned"`.
  planned: boolean;
  coverage: HorizonCoverage;
  is_current: boolean;
  running_distance_m: number | null;
  effort_score: number | null;
  // discipline/intent -> share of the week's load, 0..1. Shares, not absolutes,
  // so a mix can never contradict the total it is a mix of. An all-zero week
  // yields an empty map rather than a fake even split.
  discipline_mix: Record<string, number>;
  intent_mix: Record<string, number>;
}

export interface ScheduleHorizon {
  weeks: HorizonWeek[];
  races: GoalRace[];
  has_plan: boolean;
  // The largest weekly load in the window. Bar lengths are true proportions of
  // it — null (or zero) means there is nothing to scale against and every week
  // draws empty.
  peak_effort_score: number | null;
}
